Compiler hangs compiling property wrapper instance with throwing default init

**Describe the bug**
This simple program hangs `swift-frontend`. (It also renders Xcode playgrounds unusable until relaunching Xcode!)

**Steps To Reproduce**
Compile this program (`swiftc main.swift`):
```swift
// main.swift
@propertyWrapper
struct Wrapper {
    var wrappedValue: Int?
    init() throws {}
}

struct Repro {
    @Wrapper var x
}
```

**Expected behavior**
Expected the compiler to produce a diagnostic like it did with Xcode 13.4 / Swift 5.6:
```
main.swift:9:6: error: call can throw, but errors cannot be thrown out of a property initializer
    @Wrapper var x
     ^
```

Actual result: Swift 5.7 compiler hangs; see sampled stack trace of `swift-frontend` below.

**Environment (please fill out the following information)**
 - OS: macOS 12.6
 - Xcode Version/Tag/Branch:
   - Xcode 14.0.1 \
     swift-driver version: 1.62.8 Apple Swift version 5.7 (swiftlang-5.7.0.127.4 clang-1400.0.29.50) \
     Target: x86_64-apple-macosx12.0
   - Xcode 14.1 beta 3 \
     swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.134.3 clang-1400.0.29.51) \
     Target: x86_64-apple-macosx12.0

**Additional context**
```
Sampling process 18868 for 3 seconds with 1 millisecond of run time between samples
Sampling completed, processing symbols...
Analysis of sampling swift-frontend (pid 18868) every 1 millisecond
Process:         swift-frontend [18868]
Path:            /Applications/Xcode-14.0.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend
Load Address:    0x109432000
Identifier:      swift-frontend
Version:         5.7.0.127.4
Code Type:       X86-64
Platform:        macOS
Parent Process:  swift-driver [18864]

Date/Time:       2022-09-29 23:35:49.571 +0300
Launch Time:     2022-09-29 23:35:39.698 +0300
OS Version:      macOS 12.6 (21G115)
Report Version:  7
Analysis Tool:   /usr/bin/sample

Physical footprint:         13.3M
Physical footprint (peak):  13.3M
----

Call graph:
    2317 Thread_2224183   DispatchQueue_1: com.apple.main-thread  (serial)
      2317 start  (in dyld) + 462  [0x115ecf52e]
        2317 swift::mainEntry(int, char const**)  (in swift-frontend) + 3082  [0x1095518ea]
          2317 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*)  (in swift-frontend) + 5608  [0x1095bb958]
            2317 llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(swift::ASTLoweringRequest const&)  (in swift-frontend) + 653  [0x109c2916d]
              2317 swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::ASTLoweringRequest const&, swift::Evaluator&)  (in swift-frontend) + 195  [0x109d17ea3]
                2317 swift::ASTLoweringRequest::evaluate(swift::Evaluator&, swift::ASTLoweringDescriptor) const  (in swift-frontend) + 4390  [0x109c24fb6]
                  2317 swift::ASTVisitor<swift::Lowering::SILGenModule, void, void, void, void, void, void>::visit(swift::Decl*)  (in swift-frontend) + 82  [0x109c26f12]
                    2317 (anonymous namespace)::SILGenType::emitType()  (in swift-frontend) + 4339  [0x109d291a3]
                      2317 emitOrDelayFunction(swift::Lowering::SILGenModule&, swift::SILDeclRef, bool)  (in swift-frontend) + 217  [0x109c220e9]
                        2317 swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*)  (in swift-frontend) + 14264  [0x109c20468]
                          2317 swift::Lowering::SILGenFunction::emitGeneratorFunction(swift::SILDeclRef, swift::Expr*, bool)  (in swift-frontend) + 2085  [0x109ccb165]
                            2317 swift::Lowering::SILGenFunction::emitReturnExpr(swift::SILLocation, swift::Expr*)  (in swift-frontend) + 1027  [0x109d1eb73]
                              2317 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext)  (in swift-frontend) + 19119  [0x109cad34f]
                                2317 swift::Lowering::SILGenFunction::emitApplyExpr(swift::ApplyExpr*, swift::Lowering::SGFContext)  (in swift-frontend) + 3272  [0x109c334e8]
                                  2317 (anonymous namespace)::CallEmission::apply(swift::Lowering::SGFContext)  (in swift-frontend) + 5125  [0x109c36e15]
                                    2317 swift::Lowering::SILGenFunction::emitApply(std::__1::unique_ptr<swift::Lowering::ResultPlan, std::__1::default_delete<swift::Lowering::ResultPlan> >&&, swift::Lowering::ArgumentScope&&, swift::SILLocation, swift::Lowering::ManagedValue, swift::SubstitutionMap, llvm::ArrayRef<swift::Lowering::ManagedValue>, swift::Lowering::CalleeTypeInfo const&, swift::OptionSet<swift::ApplyFlags, unsigned char>, swift::Lowering::SGFContext, llvm::Optional<swift::ImplicitActorHopTarget>)  (in swift-frontend) + 846  [0x109c2e33e]
                                      2317 emitRawApply(swift::Lowering::SILGenFunction&, swift::SILLocation, swift::Lowering::ManagedValue, swift::SubstitutionMap, llvm::ArrayRef<swift::Lowering::ManagedValue>, swift::CanTypeWrapper<swift::SILFunctionType>, swift::OptionSet<swift::ApplyFlags, unsigned char>, llvm::ArrayRef<swift::SILValue>, llvm::SmallVectorImpl<swift::SILValue>&, swift::Lowering::ExecutorBreadcrumb)  (in swift-frontend) + 1583  [0x109c2dd8f]
                                        2317 swift::Lowering::SILGenFunction::getTryApplyErrorDest(swift::SILLocation, swift::CanTypeWrapper<swift::SILFunctionType>, swift::Lowering::ExecutorBreadcrumb, swift::SILResultInfo, bool)  (in swift-frontend) + 645  [0x109d1f205]
                                          2317 swift::Lowering::CleanupManager::emitBranchAndCleanups(swift::Lowering::JumpDest, swift::SILLocation, llvm::ArrayRef<swift::SILValue>, swift::Lowering::ForUnwind_t)  (in swift-frontend) + 54  [0x109c0bf46]
                                            2317 swift::Lowering::CleanupManager::emitCleanups(swift::DiverseStackBase::stable_iterator, swift::CleanupLocation, swift::Lowering::ForUnwind_t, bool)  (in swift-frontend) + 442,451,...  [0x109c0be0a,0x109c0be13,...]

Total number in stack (recursive counted multiple, when >=5):

Sort by top of stack, same collapsed (when >= 5):
        swift::Lowering::CleanupManager::emitCleanups(swift::DiverseStackBase::stable_iterator, swift::CleanupLocation, swift::Lowering::ForUnwind_t, bool)  (in swift-frontend)        2317
```

--- a/lib/Sema/CSGen.cpp
+++ b/lib/Sema/CSGen.cpp
@@ -4043,10 +4043,10 @@ generateForEachStmtConstraints(
   // `.next()`. `next()` is called on each iteration of the loop.
   {
     auto *nextRef = UnresolvedDotExpr::createImplicit(
-      ctx,
-      new (ctx) DeclRefExpr(makeIteratorVar, DeclNameLoc(),
-                            /*Implicit=*/true),
-      ctx.Id_next, /*labels=*/ArrayRef<Identifier>());
+        ctx,
+        new (ctx) DeclRefExpr(makeIteratorVar, DeclNameLoc(stmt->getForLoc()),
+                              /*Implicit=*/true),
+        ctx.Id_next, /*labels=*/ArrayRef<Identifier>());
     nextRef->setFunctionRefKind(FunctionRefKind::SingleApply);
 
     Expr *nextCall = CallExpr::createImplicitEmpty(ctx, nextRef);

--- a/lib/Sema/TypeCheckEffects.cpp
+++ b/lib/Sema/TypeCheckEffects.cpp
@@ -1775,9 +1775,6 @@ public:
   void diagnoseUnhandledThrowSite(DiagnosticEngine &Diags, ASTNode E,
                                   bool isTryCovered,
                                   const PotentialEffectReason &reason) {
-    if (E.isImplicit())
-      return;
-
     switch (getKind()) {
     case Kind::PotentiallyHandled:
       if (IsNonExhaustiveCatch) {

--- a/test/Concurrency/async_sequence_syntax.swift
+++ b/test/Concurrency/async_sequence_syntax.swift
@@ -9,7 +9,10 @@ func missingAsync<T : AsyncSequence>(_ seq: T) throws {
 
 @available(SwiftStdlib 5.1, *)
 func missingThrows<T : AsyncSequence>(_ seq: T) async {
-  for try await _ in seq { } // expected-error{{error is not handled because the enclosing function is not declared 'throws'}}
+  for try await _ in seq { } 
+  // expected-error@-1 {{error is not handled because the enclosing function is not declared 'throws'}}
+  // expected-error@-2 {{call can throw, but the error is not handled}}
+  // expected-note@-3 {{call is to 'rethrows' function, but a conformance has a throwing witness}}
 }
 
 @available(SwiftStdlib 5.1, *)
@@ -27,7 +30,9 @@ func missingThrowingInBlock<T : AsyncSequence>(_ seq: T) {
 @available(SwiftStdlib 5.1, *)
 func missingTryInBlock<T : AsyncSequence>(_ seq: T) { 
   executeAsync { 
-    for await _ in seq { } // expected-error{{call can throw, but the error is not handled}}
+    for await _ in seq { } 
+    // expected-error@-1 2{{call can throw, but the error is not handled}}
+    // expected-note@-2 {{call is to 'rethrows' function, but a conformance has a throwing witness}}
   }
 }
 

--- a/test/decl/func/throwing_functions.swift
+++ b/test/decl/func/throwing_functions.swift
@@ -279,3 +279,14 @@ struct FunctionHolder {
   }
 }
 
+// https://github.com/apple/swift/issues/61368
+
+@propertyWrapper
+struct Wrapper {
+  var wrappedValue: Int?
+  init() throws {}
+}
+
+struct Repro {
+  @Wrapper var x // expected-error {{call can throw, but errors cannot be thrown out of a property initializer}}
+}

